library(vcfR)
library(adegenet)
library(ggplot2)
library(RColorBrewer)
library(grDevices)

################################### In remote supercomputer (M3 cluster) #######################################

#vcf@gt[vcf@gt == "./.:.:.:.:."] <- NA
#The above is after using vcftools, which changes the format of missing data, stopping read.vcfR to convert it to NAs

# Read vcf into R
# This file has already being filetered for allele frequency (<0.01) in stacks
vcf <- read.vcfR("populations.snps.vcf")

# Filter loci for missing data
miss <- apply(dp, MARGIN = 1, function(x){sum(is.na(x))})
miss <- miss/ncol(dp)
vcf <- vcf[miss < 0.6, ]
vcf

write.vcf(vcf, "filteredSNPs_ML0.6.vcf")


######################## In local PC #####################################

vcf <- read.vcfR("C:/Users/Cris/OneDrive/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Sequenced data/Stacks/pop_output/filteredSNPs_ML0.6.vcf")
vcf

# Filter vcf by depth <5
dp <- extract.gt(vcf, element = "DP", as.numeric = TRUE)
summary(as.vector(dp))
index <- cbind(rep(FALSE, nrow(vcf@gt)), dp < 5)
vcf@gt[index] <- NA
dp <- extract.gt(vcf, element = "DP", as.numeric = TRUE)
summary(as.vector(dp))
vcf

# Filter vcf by genotype quality <30
gq <- extract.gt(vcf, element = "GQ", as.numeric = TRUE)
summary(as.vector(gq))
index <- cbind(rep(FALSE, nrow(vcf@gt)), gq < 30)
vcf@gt[index] <- NA
gq <- extract.gt(vcf, element = "GQ", as.numeric = TRUE)
summary(as.vector(gq))
vcf

# Filter samples for missing loci
miss <- apply(dp, MARGIN = 2, function(x){sum(is.na(x))})
miss <- miss/nrow(dp)
vcf@gt <- vcf@gt[ , c(TRUE, miss < 0.6)]
vcf

# Filter loci for missing data
miss <- apply(dp, MARGIN = 1, function(x){sum(is.na(x))})
miss <- miss/ncol(dp)
vcf <- vcf[miss < 0.55, ]
vcf

## Filter for Linkage disequilibrium
library(gaston)
file <- "C:/Users/Cris/OneDrive/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Sequenced data/Stacks/pop_output/filteredSNPs_AF0.01-DP5-GQ30-ML0.6-MD0.55.vcf"
gal.vcf <- read.vcfR(file)
galbedmatrix <- read.vcf(file)
gal.ld <- LD(galbedmatrix, lim = c(1, ncol(galbedmatrix)), measure = "r2")
LD.plot(gal.ld[1:20,1:20], snp.positions = galbedmatrix@snps$pos[1:20])
ldvector <- LD.thin(galbedmatrix, threshold = 0.80, max.dist = 500e3, beg = 1, end = ncol(galbedmatrix),
                    dist.unit = "bases", extract = FALSE)
sum(ldvector)
gal.vcf <- gal.vcf[ldvector, ]
gal.vcf

# Write filtered vcf (large SNP set)
file <- "C:/Users/Cris/OneDrive/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Sequenced data/Stacks/pop_output/large_SNPs.vcf.gz"
write.vcf(gal.vcf, file)

# Filter loci for missing data (2)
miss <- apply(dp, MARGIN = 1, function(x){sum(is.na(x))})
miss <- miss/ncol(dp)
gal.vcf <- gal.vcf[miss < 0.3, ]
gal.vcf

# Write filtered vcf (reduced SNP set)
file <- "C:/Users/Cris/OneDrive/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Sequenced data/Stacks/pop_output/reduced_SNPs.vcf.gz"
write.vcf(gal.vcf, file)

## done