# Modify vcf files for plink
library(vcfR)
file <- "C:/Users/Cris/OneDrive/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Sequenced data/Stacks/pop_output/reduced_SNPs.vcf.gz"
vcf <- read.vcfR(file)
vcf@fix[, 1] <- paste0("X", vcf@fix[, 1])
write.vcf(vcf, "C:/Users/Cris/OneDrive/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Sequenced data/SNPs/ADMIXTURE/vcf_plink.vcf.gz")

# Read admixture output
file <- "/Users/cris/Library/CloudStorage/OneDrive-Personal/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Sequenced data/SNPs/ADMIXTURE/galplink.2.Q"
admixq <- read.table(file)
barplot(t(admixq), col = c("blue4", "red3"), border = NA, space = 0, xlab = "Individuals", ylab = "Admixture coefficients")
qmatrix <- admixq

# read in pop map and filter for missing individuals
file <- "/Users/cris/Library/CloudStorage/OneDrive-Personal/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Sequenced data/Stacks/pop_output/reduced_SNPs.vcf.gz"
gal.vcf <- read.vcfR(file)
gal.gl <- vcfR2genlight(gal.vcf)
gal.gl
file <- "/Users/cris/Library/CloudStorage/OneDrive-Personal/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Sequenced data/Stacks/popmap.txt"
popmap <- read.table(file)
n <- length(gal.gl$gen) + 1
samples <- names(gal.vcf@gt[1, 2:n])
popdata <- popmap[popmap$V1 %in% samples, ]
pop(gal.gl) <- popdata[ , 2]

# Generate coordinates file
library(xlsx)
file <- "/Users/cris/Library/CloudStorage/OneDrive-Personal/Documents/Monash University/MonroLab/Sampling Sites.xlsx"
coordinates <- read.xlsx(file, sheetIndex = 1, rowIndex = 1:31, colIndex = 4:6)
coordinates <- merge(coordinates, popdata, by.x = "Full_ID", by.y = "V2")
coordinates <- coordinates[ , c(4, 1, 2, 3)]
names(coordinates) <- c("sample", "pop", "lat", "long")
coords <- unname(coordinates[ , 3:4])
#

# Plot structure results with ggplpot
# *Some of this code was taken from the structurePlot function in strataG package
library(ggplot2)
library(tidyr)
admixturedata <- data.frame(sample = coordinates$sample, pop = coordinates$pop, k1 = qmatrix[ , 1], k2 = qmatrix[ , 2])
pop.freq <- table(admixturedata$pop)
admixturedata$pop <- as.factor(admixturedata$pop)
levels(admixturedata$pop) <- paste(levels(admixturedata$pop), "\n(n = ", pop.freq, ")", sep = "")
admixturedata$sample <- 1:nrow(admixturedata)
names(admixturedata)[1] <- "x"
pop.cntr <- tapply(admixturedata$x, admixturedata$pop, mean)
pop.div <- tapply(admixturedata$x, admixturedata$pop, min)[-1] - 0.5
admixturedata <- gather(admixturedata, "k_group", "ad_coef", 3:4)
names(admixturedata) <- c("x", "pop", "group", "adcoef")

segments <- data.frame(
  x = unname(pop.div),
  xend = unname(pop.div),
  y = rep(-0.03, length(unname(pop.div))),
  yend = rep(1, length(unname(pop.div)))
)

png("C:/Users/Cris/OneDrive/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Figures/gal_admixture.png", width = 2500, height = 500)
gg <- ggplot(admixturedata, aes_string("x", "adcoef"))
gg +
  geom_bar(aes_string(fill = "group"), stat = "identity", width = 0.9) + 
  labs(y = "Admixture coefficients", x = "Individuals") + 
  theme(axis.ticks.x = element_blank(), legend.position = "none", legend.title = element_blank(), panel.background = element_blank(),
        axis.line.y = element_line(colour = "grey", size = 1.0), axis.line.x = element_line(colour = "grey", size = 1.0),
        axis.title = element_text(size = 30), axis.text.y = element_text(size = 20), axis.text.x = element_text(size = 12)) +
  geom_segment(data = segments, aes(x = x, y = y, xend = xend, yend = yend), size = 1) +
  scale_x_continuous(name = "Individuals", breaks = pop.cntr, labels = names(pop.cntr), expand = c(0.01, 0.01)) +
  scale_fill_manual(values = c("steelblue1", "tomato2"))
dev.off()

# Same plot formatted for manuscript
library(stringr)
xlabels = str_split(as.character(unique(coordinates$pop)), "_", simplify = TRUE)[ , 1]
segments$y <- 0
pop.div <- c(0, unname(tapply(admixturedata$x, admixturedata$pop, min)[-1]) - 1, length(gal.gl@ind.names))
segmentshor <- segments[1:30, ]
segmentshor <- rbind(segments, c(NA, NA))
segmentshor$y <- -0.01
segmentshor$yend <- -0.01
segmentshor$x <- (pop.div+0.6)[1:length(pop.div)-1]
segmentshor$xend <- (pop.div+0.4)[2:length(pop.div)]

png("C:/Users/Cris/OneDrive/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Figures/gal_admixture_formatted.png", width = 2500, height = 500)
gg <- ggplot(admixturedata, aes_string("x", "adcoef"))
gg +
  geom_bar(aes_string(fill = "group"), stat = "identity", width = 0.9) + 
  labs(y = "", x = "") + 
  theme(axis.ticks = element_blank(), legend.position = "none", legend.title = element_blank(), panel.background = element_blank(),
        axis.line.y = element_blank(), axis.line.x = element_blank(),
        axis.title = element_blank(), axis.text.y = element_blank(), 
        axis.text.x = element_text(size = 20, angle = c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
                                   vjust = c(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1))) +
  geom_segment(data = segments, aes(x = x, y = y, xend = xend, yend = yend), size = 0.6) +
  geom_segment(data = segmentshor, aes(x = x, y = y, xend = xend, yend = yend), size = 1) +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(breaks = pop.cntr, labels = xlabels, expand = c(0, 0)) +
  scale_fill_manual(values = c("steelblue1", "tomato2"))
dev.off()
# 2500w x 500h
##

# Plot without pop names

index <- rep(c(FALSE, TRUE), 15)
segmentshor[index, 3:4] <- segmentshor[index, 3:4] - 0.01


# png("C:/Users/Cris/OneDrive/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Figures/gal_admixture_formatted2.png", width = 2500, height = 500)
pdf("/Users/cris/Library/CloudStorage/OneDrive-Personal/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Figures/gal_admixture_formatted3.pdf", width = 25, height = 5)
gg <- ggplot(admixturedata, aes_string("x", "adcoef"))
gg <- gg +
  geom_bar(aes_string(fill = "group"), stat = "identity", width = 0.8) + 
  labs(y = "", x = "") + 
  theme(axis.ticks = element_blank(), legend.position = "none", legend.title = element_blank(), panel.background = element_blank(),
        axis.line.y = element_blank(), axis.line.x = element_blank(),
        axis.title = element_blank(), axis.text.y = element_blank(), 
        axis.text.x = element_blank()) +
  geom_segment(data = segments, aes(x = x, y = y, xend = xend, yend = yend), size = 0.6) +
  # geom_segment(data = segmentshor, aes(x = x, y = y, xend = xend, yend = yend), size = 1) +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(breaks = pop.cntr, labels = xlabels, expand = c(0, 0)) +
  scale_fill_manual(values = c("steelblue1", "tomato2"))
gg
dev.off()

ggpreview <- function(...) {
  fname <- tempfile(fileext = ".png")
  ggsave(filename = fname, ...)
  system2("open", fname)
  invisible(NULL)
}

ggpreview(scale = 1, width = 25, height = 8, units = "cm", limitsize = F)

ggsave("/Users/cris/Library/CloudStorage/OneDrive-Personal/Documents/Monash University/MonroLab/Chapter 2/Re-sequencing/Figures/gal_admixture_formatted3.pdf"
       , plot = gg
       , device = "pdf"
       , scale = 1
       , width = 25, height = 6
       , units = "cm"
       , limitsize = F)
