### I. Prepare inputs ###

# Prepare temperature variables

# Scale variables (u = 0, sd = 1)
file <- "~/temp_variables.csv"
tempvardf <- read.csv(file)
for(i in 4:ncol(tempvardf)){
  tempvardf[ , i] <- tempvardf[ , i] - mean(tempvardf[ , i])
  tempvardf[ , i] <- tempvardf[ , i]/sd(tempvardf[ , i])
}
#

# Correlation matrix to identify and remove highly correlated variables
library(Hmisc)
library(corrplot)
cormat <- rcorr(as.matrix(tempvardf[ , 4:ncol(tempvardf)]))
corrplot(cormat$r, type = "upper", tl.col = "black", tl.srt = 45)
# Variables to keep: gm (trend), mmr (fluctuation), and maxwm (extreme), noisecol (noise)
filtempvar <- tempvardf[ , c(1:4, 6, 9, 15)]
file <- "~/temp_variables(filt.and.scaled).csv"
write.csv(filtempvar, file, row.names = FALSE)
#

# Create covariate data file (with temperature variables) for BayPass
file <- "~/temp_variables(filt.and.scaled).csv"
filtempvar <- read.csv(file)

# caes covariates
spSplit <- read.csv("~/species_split.csv")
caespops <- unique(spSplit$pop[spSplit$caes])
caesvar <- filtempvar[filtempvar$pop %in% caespops, 4:7]
caesvar <- t(caesvar)
colnames(caesvar) <- caespops
file <- "~/caescovariates.txt"
write.table(caesvar, file, row.names = FALSE, col.names = FALSE)
file <- "~/caescovariates(baypass).csv"
write.csv(caesvar, file, row.names = TRUE)
#

# gem covariates
spSplit <- read.csv("~/species_split.csv")
gempops <- unique(spSplit$pop[spSplit$gem])
gemvar <- filtempvar[filtempvar$pop %in% gempops, 4:7]
gemvar <- t(gemvar)
colnames(gemvar) <- gempops
file <- "~/gemcovariates.txt"
write.table(gemvar, file, row.names = FALSE, col.names = FALSE)
file <- "~/gemcovariates(baypass).csv"
write.csv(gemvar, file, row.names = TRUE)
#


# Prepare SNP matrix

library(vcfR)
# Read one of these at a time
file <- "~/SNPs_caes_BIG.vcf"
file <- "~/SNPs_caes_SMALL.vcf"
file <- "~/SNPs_gem_BIG.vcf"
file <- "~/SNPs_gem_SMALL.vcf"
vcf <- read.vcfR(file)

# Convert to genlight
gl <- vcfR2genlight(vcf)
gl

# read in pop map and filter for missing individuals
file <- "~/popmap.txt"
popmap <- read.table(file)
n <- length(gl$gen) + 1
samples <- names(vcf@gt[1, 2:n])
popdata <- popmap[popmap$V1 %in% samples, ]
pop(gl) <- popdata[ , 2]

# Convert to genepop format
library(radiator)
genepop <- genomic_converter(gl, output = "genepop")

# Read in genepop file and add pop data
file <- "-5_radiator_genomic_converter_20210105@1757/radiator_data_20210105@1757_genepop.gen" # Read file corresponding to each species X SNP set combination
genepop <- my.read.genepop(file)
genepop$popinfo <- as.character(popdata[ , 2])

# Convert to 0 (homozygote for non-focal allele), 1 (heterozygote), 2 (homozygote for focal allele) format and save (this is the most standard and easier to work with format)
SNPmat <- data.frame(lapply(genepop, as.character), stringsAsFactors = FALSE)
for(i in 3:ncol(SNPmat)){
  for(j in 1:nrow(SNPmat)){
    SNPmat[j, i] <- if(SNPmat[j, i] == "001001"){0} else if(SNPmat[j, i] == "001002"){1} else if(SNPmat[j, i] == "002002"){2} else {NA}
  }
}
sp <- "caes" # caes/gem
size <- "BIG" # BIG/SMALL
write.csv(SNPmat, paste0("SNPmatrix_", sp, size, ".csv"))

# Add pop column to SNPmat
file <- "~/popmap.txt"
popmap <- read.table(file)
n <- nrow(SNPmat)
samples <- row.names(SNPmat)
popdata <- popmap[popmap$V1 %in% samples, ]
SNPmat <- cbind(pop = popdata$V2, SNPmat)

# Convert SNPmat to baypass (bp)
pops <- unique(SNPmat$pop)
SNPbp <- data.frame()
for(i in 2:ncol(SNPmat)){
  vec <- vector(mode = "integer")
  for(j in 1:length(pops)){
    alleles <- as.numeric(SNPmat[SNPmat$pop == pops[j], i])
    alleles[alleles == -1] <- NA
    no_ind <- length(na.omit(alleles))
    count2 <- sum(alleles, na.rm = TRUE)
    count1 <- no_ind*2 - count2
    vec <- c(vec, count1, count2)
  }
  SNPbp <- rbind(SNPbp, vec)
}
sum(is.na(SNPbp))
row.names(SNPbp) <- names(SNPmat)[-1]
names(SNPbp) <- sort(rep(pops, 2))

# Write csv and txt files
sp <- "caes" # caes/gem
size <- "BIG" # BIG/SMALL
file <- paste0("~/SNPmat_", sp, "_", size, "_(baypass).csv")
write.csv(SNPbp, file)
file <- paste0("~/SNPmat_", sp, "_", size, ".txt")
write.table(SNPbp, file, row.names = FALSE, col.names = FALSE)



### II. Read and analyse outputs ###

#########################  G. caespitosa ###############################
# Read baypass outputs
file <- "bp_stdcov_LDfil_caes_summary_pi_xtx.out"
caesxtx <- read.table(file, header = TRUE)
file <- "bp_stdcov_LDfil_caes_summary_betai_reg.out"
caesBF <- read.table(file, header = TRUE)

# Total number of significant XtX outliers
plot(caesxtx$XtXst)
# log10(1/0.05) # = 1.30103
XtX_cutoff <- min(caesxtx$XtXst[which(caesxtx$log10.1.pval. > 1.30103)]) # Outliers have XtX >= XtX_cutoff
abline(h = XtX_cutoff, col = "blue")
sum(caesxtx$log10.1.pval. > 1.30103) 
# 1519 outlier loci

# Number of SNPs significantly associated to temperature variables and XtX outliers
plot(caesBF$BF.dB.)
abline(h = 10, col = "blue") # 10<BF<15 => "strong evidence"
abline(h = 15, col = "blue") # 15<BF<20 => "very strong evidence"
abline(h = 20, col = "blue") # 20<BF => "decisive evidence"

BF_cutoff <- 10

plot(caesxtx$XtXst, caesBF[caesBF$COVARIABLE == 1, 5]) # covaribale 1 => gm (mean)
abline(v = XtX_cutoff, col = "blue")
abline(h = BF_cutoff, col = "blue")
sum(caesBF[caesBF$COVARIABLE == 1, 5] > BF_cutoff) # 953 SNPs significantlly associated to gm
sum(caesBF[caesBF$COVARIABLE == 1, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff) # 102 SNPs significantlly associated to gm and XtX outlier

plot(caesxtx$XtXst, caesBF[caesBF$COVARIABLE == 2, 5]) # covaribale 2 => mmr (mean monthly range)
abline(v = XtX_cutoff, col = "blue")
abline(h = BF_cutoff, col = "blue")
sum(caesBF[caesBF$COVARIABLE == 2, 5] > BF_cutoff) # 97 SNPs significantlly associated to mmr
sum(caesBF[caesBF$COVARIABLE == 2, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff) # 8 SNPs significantlly associated to mmr and XtX outlier

plot(caesxtx$XtXst, caesBF[caesBF$COVARIABLE == 3, 5]) # covaribale 3 => maxwm (max temperature of the warmest month)
abline(v = XtX_cutoff, col = "blue")
abline(h = BF_cutoff, col = "blue")
sum(caesBF[caesBF$COVARIABLE == 3, 5] > BF_cutoff) # 111 SNPs significantlly associated to maxwm
sum(caesBF[caesBF$COVARIABLE == 3, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff) # 11 SNPs significantlly associated to maxwm and XtX outlier

plot(caesxtx$XtXst, caesBF[caesBF$COVARIABLE == 4, 5]) # covaribale 4 => noisecol (noise colour)
abline(v = XtX_cutoff, col = "blue")
abline(h = BF_cutoff, col = "blue")
sum(caesBF[caesBF$COVARIABLE == 4, 5] > BF_cutoff) # 51 SNPs significantlly associated to noisecol
sum(caesBF[caesBF$COVARIABLE == 4, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff) # 4 SNPs significantlly associated to noisecol and XtX outlier

# Total candidate SNPs
sum(caesBF[caesBF$COVARIABLE == 1, 5] > BF_cutoff | caesBF[caesBF$COVARIABLE == 2, 5] > BF_cutoff | caesBF[caesBF$COVARIABLE == 3, 5] > BF_cutoff | caesBF[caesBF$COVARIABLE == 4, 5] > BF_cutoff) # 1183 SNPs associated to a temperature variable in total
sum(caesBF[caesBF$COVARIABLE == 1, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff |
      caesBF[caesBF$COVARIABLE == 2, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff |
      caesBF[caesBF$COVARIABLE == 3, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff |
      caesBF[caesBF$COVARIABLE == 4, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff) # 119 SNPs associated to a temperature variable and XtX outlier in total

# Extract candidate SNPs IDs
snpmat.caes <- read.csv("SNPmat_caes_BIG.csv", row.names = 1)
caesxtx$MRK <- row.names(snpmat.caes)
caesBF$MRK <- rep(row.names(snpmat.caes), 4)
cand_caes_gm <- caesxtx$MRK[caesBF[caesBF$COVARIABLE == 1, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff]
cand_caes_mmr <- caesxtx$MRK[caesBF[caesBF$COVARIABLE == 2, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff]
cand_caes_max <- cand_gm <- caesxtx$MRK[caesBF[caesBF$COVARIABLE == 3, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff]
cand_caes_noise <- cand_gm <- caesxtx$MRK[caesBF[caesBF$COVARIABLE == 4, 5] > BF_cutoff & caesxtx$XtXst > XtX_cutoff]
cand_caes <- union(cand_caes_gm, union(cand_caes_mmr, union(cand_caes_max, cand_caes_noise)))
write(cand_caes, "candidates_caes_stdcovBF10.txt")

# regresion coefficients (strength of association) (>0.2 is strong)
plot(caesxtx$XtXst, caesBF[caesBF$COVARIABLE == 4, "M_Beta"])
abline(v = XtX_cutoff, col = "blue")
abline(h = 0.2, col = "blue")
abline(h = -0.2, col = "blue")
sum(caesBF[caesBF$COVARIABLE == 1, "M_Beta"] > 0.2 | caesBF[caesBF$COVARIABLE == 1, "M_Beta"] < -0.2) # 29 SNPs have Beta_i > 0.2 ("strong regression coefficient") for mean temp
sum((caesBF[caesBF$COVARIABLE == 1, "M_Beta"] > 0.2 | caesBF[caesBF$COVARIABLE == 1, "M_Beta"] < -0.2) & caesxtx$XtXst > XtX_cutoff) # 0 have Beta_i > 0.2 ("strong regression coefficient") for mean temp
sum(caesBF[caesBF$COVARIABLE == 2, "M_Beta"] > 0.2 | caesBF[caesBF$COVARIABLE == 2, "M_Beta"] < -0.2) # 7 SNPs have Beta_i > 0.2 ("strong regression coefficient") for mean monthly range
sum((caesBF[caesBF$COVARIABLE == 2, "M_Beta"] > 0.2 | caesBF[caesBF$COVARIABLE == 2, "M_Beta"] < -0.2) & caesxtx$XtXst > XtX_cutoff) # 0 have Beta_i > 0.2 ("strong regression coefficient") for mean monthly range
sum(caesBF[caesBF$COVARIABLE == 3, "M_Beta"] > 0.2 | caesBF[caesBF$COVARIABLE == 3, "M_Beta"] < -0.2) # 6 SNPs have Beta_i > 0.2 ("strong regression coefficient") for max temp
sum((caesBF[caesBF$COVARIABLE == 3, "M_Beta"] > 0.2 | caesBF[caesBF$COVARIABLE == 3, "M_Beta"] < -0.2) & caesxtx$XtXst > XtX_cutoff) # 2 have Beta_i > 0.2 ("strong regression coefficient") for max temp
sum(caesBF[caesBF$COVARIABLE == 4, "M_Beta"] > 0.2 | caesBF[caesBF$COVARIABLE == 4, "M_Beta"] < -0.2) # 4 SNPs have Beta_i > 0.2 ("strong regression coefficient") for noise colour
sum((caesBF[caesBF$COVARIABLE == 4, "M_Beta"] > 0.2 | caesBF[caesBF$COVARIABLE == 4, "M_Beta"] < -0.2) & caesxtx$XtXst > XtX_cutoff) # 1 have Beta_i > 0.2 ("strong regression coefficient") for noise colour

##### Nice plot for manuscript ######
library(ggplot2)
# handy preview file for ggsave() (from here: https://gist.github.com/tjmahr/1dd36d78ecb3cff10baf01817a56e895)
ggpreview <- function(...) {
  fname <- tempfile(fileext = ".png")
  ggsave(filename = fname, ...)
  system2("open", fname)
  invisible(NULL)
}

# Pie chart with distribution of outliers among variables
caes.gm <- sum(caesBF[caesBF$COVARIABLE == 1, 5] > 10 & caesxtx$XtXst > 31.4)
caes.mmr <- sum(caesBF[caesBF$COVARIABLE == 2, 5] > 10 & caesxtx$XtXst > 31.4)
caes.maxwm <- sum(caesBF[caesBF$COVARIABLE == 3, 5] > 10 & caesxtx$XtXst > 31.4)
caes.noisecol <- sum(caesBF[caesBF$COVARIABLE == 4, 5] > 10 & caesxtx$XtXst > 31.4)
df <- data.frame(
  Temperature = c("Mean", "Maximum", "Range", "Noise"),
  value = c(caes.gm, caes.maxwm, caes.mmr, caes.noisecol)
)
caes.pie <- ggplot(df, aes(x = "", y = value, fill = Temperature, alpha = 0.5)) + 
  geom_col(color = "black") +
  # geom_bar(width = 1, stat = "identity", alpha = 0.5) +
  coord_polar("y", start = 0) +
  theme_void() +
  theme(legend.position="none") +
  scale_fill_manual(values = c("yellow", "red", "blue", "green"))
caes.pie
caes.pie.t <- ggplotGrob(caes.pie)

# BayPass plot with XtX outliers and BFs
caes.df <- data.frame(var = as.factor(c(rep("gm", 20782), rep("mmr", 20782), rep("maxwm", 20782), rep("noisecol", 20782))),
                      xtx = caesxtx$XtXst, bf = c(caesBF[caesBF$COVARIABLE == 1, 5], caesBF[caesBF$COVARIABLE == 2, 5],
                                                  caesBF[caesBF$COVARIABLE == 3, 5], caesBF[caesBF$COVARIABLE == 4, 5]))
caes.gg <- ggplot(data = caes.df, aes(x = xtx, y = bf)) +
  geom_point(aes(fill = var), shape = 21, alpha = 0.5, size = 2) +
  scale_fill_manual(values = c("red", "yellow", "green", "blue"), name = "Temperature", labels = c("Mean", "Maximum", "Range", "Noise")) +
  xlab("XtX") + 
  ylim(-21.5, 40) +
  ylab(expression("BF"[is]~(dB))) +
  theme_classic() +
  geom_vline(xintercept = 31.4, color = "black", lty = 2) +
  geom_hline(yintercept = 10, color = "black", lty = 2) +
  ggtitle(expression(italic("G. caespitosa"))) +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_rect(aes(xmin = 31.4, xmax = Inf, ymin = 10, ymax = Inf), fill = NA, colour = "red")
caes.gg  

# Add pie chart as inset
caes.gg <- caes.gg +
  annotation_custom(grob = caes.pie.t, xmin = 77, xmax = Inf, ymin = -12, ymax = -Inf)
caes.gg


#########################  G. gemineoa ###############################
file <- "bp_stdcov_LDfil_gem_summary_pi_xtx.out"
gemxtx <- read.table(file, header = TRUE)
file <- "bp_stdcov_LDfil_gem_summary_betai_reg.out"
gemBF <- read.table(file, header = TRUE)

# Total number of significant XtX outliers
# gemxtx[gemxtx$log10.1.pval. > 1.3 & gemxtx$log10.1.pval. < 1.31, ] # This is to find the top 0.1% POD threshold (log10.1.pval. > 1.3 is significant at p < 0.05)
plot(gemxtx$XtXst)
XtX_cutoff <- min(gemxtx$XtXst[which(gemxtx$log10.1.pval. > 1.30103)]) # Outliers have XtX >= 33.92583
abline(h = XtX_cutoff, col = "blue")
sum(gemxtx$log10.1.pval. > 1.30103) 
# 1260 outlier loci

# Number of SNPs significantly associated to temperature variables and XtX outliers
plot(gemBF$BF.dB.)
abline(h = 10, col = "blue") # 10<BF<15 => "strong evidence"
abline(h = 15, col = "blue") # 15<BF<20 => "very strong evidence"
abline(h = 20, col = "blue") # 20<BF => "decisive evidence"

BF_cutoff <- 10

plot(gemxtx$XtXst, gemBF[gemBF$COVARIABLE == 1, 5]) # covaribale 1 => gm (mean)
abline(v = XtX_cutoff, col = "blue")
abline(h = BF_cutoff, col = "blue")
sum(gemBF[gemBF$COVARIABLE == 1, 5] > BF_cutoff) # 135 SNPs significantlly associated to gm
sum(gemBF[gemBF$COVARIABLE == 1, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff) # 19 SNPs significantlly associated to gm and XtX outlier

plot(gemxtx$XtXst, gemBF[gemBF$COVARIABLE == 2, 5]) # covaribale 2 => mmr (mean monthly range)
abline(v = XtX_cutoff, col = "blue")
abline(h = BF_cutoff, col = "blue")
sum(gemBF[gemBF$COVARIABLE == 2, 5] > BF_cutoff) # 123 SNPs significantlly associated to mmr
sum(gemBF[gemBF$COVARIABLE == 2, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff) # 18 SNPs significantlly associated to mmr and XtX outlier

plot(gemxtx$XtXst, gemBF[gemBF$COVARIABLE == 3, 5]) # covaribale 3 => maxwm (max temperature of the warmest month)
abline(v = XtX_cutoff, col = "blue")
abline(h = BF_cutoff, col = "blue")
sum(gemBF[gemBF$COVARIABLE == 3, 5] > BF_cutoff) # 15 SNPs significantlly associated to maxwm
sum(gemBF[gemBF$COVARIABLE == 3, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff) # 3 SNPs significantlly associated to maxwm and XtX outlier

plot(gemxtx$XtXst, gemBF[gemBF$COVARIABLE == 4, 5]) # covaribale 4 => noisecol (noise colour)
abline(v = XtX_cutoff, col = "blue")
abline(h = BF_cutoff, col = "blue")
sum(gemBF[gemBF$COVARIABLE == 4, 5] > BF_cutoff) # 128 SNPs significantlly associated to noisecol
sum(gemBF[gemBF$COVARIABLE == 4, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff) # 0 SNPs significantlly associated to noisecol and XtX outlier

# Total candidate SNPs
sum(gemBF[gemBF$COVARIABLE == 1, 5] > BF_cutoff | gemBF[gemBF$COVARIABLE == 2, 5] > BF_cutoff | gemBF[gemBF$COVARIABLE == 3, 5] > BF_cutoff | gemBF[gemBF$COVARIABLE == 4, 5] > BF_cutoff) # 385 SNPs associated to a temperature variable in total
sum(gemBF[gemBF$COVARIABLE == 1, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff |
      gemBF[gemBF$COVARIABLE == 2, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff |
      gemBF[gemBF$COVARIABLE == 3, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff |
      gemBF[gemBF$COVARIABLE == 4, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff) # 40 SNPs associated to a temperature variable and XtX outlier in total

# Extract candidate SNPs IDs
snpmat.gem <- read.csv("SNPmat_gem_BIG.csv", row.names = 1)
gemxtx$MRK <- row.names(snpmat.gem)
gemBF$MRK <- rep(row.names(snpmat.gem), 4)
cand_gem_gm <- gemxtx$MRK[gemBF[gemBF$COVARIABLE == 1, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff]
cand_gem_mmr <- gemxtx$MRK[gemBF[gemBF$COVARIABLE == 2, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff]
cand_gem_max <- cand_gm <- gemxtx$MRK[gemBF[gemBF$COVARIABLE == 3, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff]
cand_gem_noise <- cand_gm <- gemxtx$MRK[gemBF[gemBF$COVARIABLE == 4, 5] > BF_cutoff & gemxtx$XtXst > XtX_cutoff]
cand_gem <- union(cand_gem_gm, union(cand_gem_mmr, union(cand_gem_max, cand_gem_noise)))
write(cand_gem, "candidates_gem_stdcovBF10.txt")

# regresion coefficients (strength of association) (>0.2 is strong)
plot(gemxtx$XtXst, gemBF[gemBF$COVARIABLE == 1, "M_Beta"])
abline(v = XtX_cutoff, col = "blue")
abline(h = 0.2, col = "blue")
abline(h = -0.2, col = "blue")
sum(gemBF[gemBF$COVARIABLE == 1, "M_Beta"] > 0.2 | gemBF[gemBF$COVARIABLE == 1, "M_Beta"] < -0.2) # 19 SNPs have Beta_i > 0.2 ("strong regression coefficient") for mean temp
sum((gemBF[gemBF$COVARIABLE == 1, "M_Beta"] > 0.2 | gemBF[gemBF$COVARIABLE == 1, "M_Beta"] < -0.2) & gemxtx$XtXst > XtX_cutoff) # 3 have Beta_i > 0.2 ("strong regression coefficient") for mean temp
sum(gemBF[gemBF$COVARIABLE == 2, "M_Beta"] > 0.2 | gemBF[gemBF$COVARIABLE == 2, "M_Beta"] < -0.2) # 5 SNPs have Beta_i > 0.2 ("strong regression coefficient") for mean monthly range
sum((gemBF[gemBF$COVARIABLE == 2, "M_Beta"] > 0.2 | gemBF[gemBF$COVARIABLE == 2, "M_Beta"] < -0.2) & gemxtx$XtXst > XtX_cutoff) # 1 have Beta_i > 0.2 ("strong regression coefficient") for mean monthly range
sum(gemBF[gemBF$COVARIABLE == 3, "M_Beta"] > 0.2 | gemBF[gemBF$COVARIABLE == 3, "M_Beta"] < -0.2) # 1 SNPs have Beta_i > 0.2 ("strong regression coefficient") for max temp
sum((gemBF[gemBF$COVARIABLE == 3, "M_Beta"] > 0.2 | gemBF[gemBF$COVARIABLE == 3, "M_Beta"] < -0.2) & gemxtx$XtXst > XtX_cutoff) # 0 have Beta_i > 0.2 ("strong regression coefficient") for max temp
sum(gemBF[gemBF$COVARIABLE == 4, "M_Beta"] > 0.2 | gemBF[gemBF$COVARIABLE == 4, "M_Beta"] < -0.2) # 4 SNPs have Beta_i > 0.2 ("strong regression coefficient") for noise colour
sum((gemBF[gemBF$COVARIABLE == 4, "M_Beta"] > 0.2 | gemBF[gemBF$COVARIABLE == 4, "M_Beta"] < -0.2) & gemxtx$XtXst > XtX_cutoff) # 0 have Beta_i > 0.2 ("strong regression coefficient") for noise colour

# Overlap between species
length(intersect(cand_caes, cand_gem))

# Evaluate significance of number of candidate loci overlap between G. caespitosa and G. gemineoa for BayPass
# Hypergeometric test to evaluate if the overlap between two lists of elements (A and B) taken from a larger pool (C) is statistically significant 
# (larger than expected by chance).
# Can also be done online here: http://nemates.org/MA/progs/overlap_stats.html
# More info here : https://stats.stackexchange.com/questions/267/how-do-i-calculate-if-the-degree-of-overlap-between-two-lists-is-significant
# and here: https://www.biostars.org/p/8480/
nc <- 24263 #total number of SNPs screened
na <- 119 #number of candidate loci identified for G. caespitosa
nb <- 40 #number of candidate loci identified for G. gemineoa
nab <- 7 #number of overlapping loci
phyper(q = nab - 1, m = na, n = nc-na, k = nb, lower.tail = FALSE)
# p = 3.528402e-17 #Exact same number as the website above

##### Nice plots for manuscript ######
library(ggplot2)

# Pie chart with distribution of outliers among variables
gem.gm <- sum(gemBF[gemBF$COVARIABLE == 1, 5] > 10 & gemxtx$XtXst > 31.4)
gem.mmr <- sum(gemBF[gemBF$COVARIABLE == 2, 5] > 10 & gemxtx$XtXst > 31.4)
gem.maxwm <- sum(gemBF[gemBF$COVARIABLE == 3, 5] > 10 & gemxtx$XtXst > 31.4)
gem.noisecol <- sum(gemBF[gemBF$COVARIABLE == 4, 5] > 10 & gemxtx$XtXst > 31.4)
df <- data.frame(
  Temperature = c("Mean", "Maximum", "Range", "Noise"),
  value = c(gem.gm, gem.maxwm, gem.mmr, gem.noisecol)
)
gem.pie <- ggplot(df, aes(x = "", y = value, fill = Temperature, alpha = 0.5)) + 
  geom_col(color = "black") +
  # geom_bar(width = 1, stat = "identity", alpha = 0.5) +
  coord_polar("y", start = 0) +
  theme_void() +
  theme(legend.position="none") +
  scale_fill_manual(values = c("yellow", "red", "blue", "green"))
gem.pie
gem.pie.t <- ggplotGrob(gem.pie)

# BayPass plot with XtX outliers and BFs
gem.df <- data.frame(var = as.factor(c(rep("gm", 20387), rep("mmr", 20387), rep("maxwm", 20387), rep("noisecol", 20387))),
                     xtx = gemxtx$XtXst, bf = c(gemBF[gemBF$COVARIABLE == 1, 5], gemBF[gemBF$COVARIABLE == 2, 5],
                                                gemBF[gemBF$COVARIABLE == 3, 5], gemBF[gemBF$COVARIABLE == 4, 5]))
gem.gg <- ggplot(data = gem.df, aes(x = xtx, y = bf)) +
  geom_point(aes(fill = var), shape = 21, alpha = 0.5, size = 2) +
  scale_fill_manual(values = c("red", "yellow", "green", "blue"), name = "Temperature", labels = c("Mean", "Maximum", "Range", "Noise")) +
  xlab("XtX") +
  ylab("") +
  ylim(-21.5, 40) +
  # ylab(expression("BF"[is]~(dB))) +
  theme_classic() +
  geom_vline(xintercept = 31.4, color = "black", lty = 2) +
  geom_hline(yintercept = 10, color = "black", lty = 2) +
  ggtitle(expression(italic("G. gemineoa"))) +
  theme(plot.title = element_text(hjust = 0.5)) +
  # xlim(0, 140) +
  # ylim(-20, 40) +
  geom_rect(aes(xmin = 31.4, xmax = Inf, ymin = 10, ymax = Inf), fill = NA, colour = "red")
gem.gg  

# Add pie chart as inset
gem.gg <- gem.gg +
  annotation_custom(grob = gem.pie.t, xmin = 122, xmax = Inf, ymin = -12, ymax = -Inf)
gem.gg

# ggpreview(scale = 1, width = 13, height = 8, units = "cm", limitsize = F)

## Combine all plots into single figure
library(ggpubr)
galbp <- ggarrange(caes.gg, gem.gg, 
                   labels = c("A", "B"), ncol = 2, nrow = 1, align = "h",
                   font.label = list(face = "plain"), vjust = 1.25, common.legend = TRUE, legend = "right")
galbp

ggpreview(scale = 1, width = 20, height = 8, units = "cm", limitsize = F)
ggsave("~/BayPass_candidates.pdf",
       plot = galbp, device = "pdf", scale = 1, width = 20, height = 8, units = "cm", limitsize = F)
ggsave("~/BayPass_candidates.jpeg",
       plot = galbp, device = "jpeg", scale = 1, width = 20, height = 8, units = "cm", limitsize = F)
