# ClimateAdaptation-and-Vulnerability
Scripts and supporting files for analyses of "Climate adaptation and vulnerability of foundation species
in a global change hotspot" Gallegos et al. 2023 (Molecular Ecology).

Analysis pipeline

1) Demultiplex and clean raw reads data (first batch)
2) Demultiplex and clean raw reads data (second batch)
3) Concat data files from both batches for each sample
*Step 4 onwards can be run from sample files in SRA repository (PRJNA820542)
4) Run stacks pipeline on small subset of samples to define parameters (m, M, and n)
5) Run stacks pipeline to call SNPs
6) Filter SNPs to get the reduced and large SNPs set (for analyses of population genetic structure and EAAs respectively)
7) Admixture
8) Split data sets between species
9) Get SST data (present and projected)
10) BayPass to identify candidate loci
11) Population structure analyses (genetic diversity statistics, PCAs, MEMGENE)
12) Gradient forest and genetic offsets


Input files

- HI.5188.002.pool_P1_P2_P3_P4_R1.fastq.gz (not provided)
- HI.5188.002.pool_P1_P2_P3_P4_R2.fastq.gz (not provided)
- Sample files in SRA repository (https://www.ncbi.nlm.nih.gov/bioproject/PRJNA820542/)
- Intermediate VCF files available on figshare (https://doi.org/10.6084/m9.figshare.21845031)
- galeolaria_barcodes.txt
- test.popmap.txt
- popmap.txt
- pop_coordinates.csv
- Sampling sites.xlsx (some extra info to pop_coordinates.csv)
